x/upgrade/keeper/migrations.go

```diff
@@ -0,0 +1,43 @@
+package keeper
+
+import (
+	"encoding/binary"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/types"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	keeper Keeper
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper Keeper) Migrator {
+	return Migrator{keeper: keeper}
+}
+
+// Migrate1to2 migrates from version 1 to 2.
+func (m Migrator) Migrate1to2(ctx sdk.Context) error {
+	return migrateDoneUpgradeKeys(ctx, m.keeper.storeKey)
+}
+
+func migrateDoneUpgradeKeys(ctx sdk.Context, storeKey storetypes.StoreKey) error {
+	store := ctx.KVStore(storeKey)
+	oldDoneStore := prefix.NewStore(store, []byte{types.DoneByte})
+	oldDoneStoreIter := oldDoneStore.Iterator(nil, nil)
+	defer oldDoneStoreIter.Close()
+
+	for ; oldDoneStoreIter.Valid(); oldDoneStoreIter.Next() {
+		oldKey := oldDoneStoreIter.Key()
+		upgradeName := string(oldKey)
+		upgradeHeight := int64(binary.BigEndian.Uint64(oldDoneStoreIter.Value()))
+		newKey := encodeDoneKey(upgradeName, upgradeHeight)
+
+		store.Set(newKey, []byte{1})
+		oldDoneStore.Delete(oldKey)
```

Comment:
I don't understand how this is working, the iterator should be taking a mutex lock out on the CacheKVStore local database. Shouldn't we buffer all the writes and deletes, and do a single one after the iterator call ends?

Reply:
Like how does this work if say, 10 upgrade handlers were in the database

Reply:
Ah nevermind, I mixed up store sorted cache and unsorted cache, ignore this